pkg/util/validation/errors.go

```diff
@@ -6,273 +6,158 @@
 
+	"github.com/grafana/mimir/pkg/distributor/distributorerror"
 	"github.com/grafana/mimir/pkg/mimirpb"
 	"github.com/grafana/mimir/pkg/util/globalerror"
 )
 
-// ValidationError is an error returned by series validation.
-//
-//nolint:revive // ignore stutter warning
-type ValidationError error
-
-// genericValidationError is a basic implementation of ValidationError which can be used when the
-// error format only contains the cause and the series.
-type genericValidationError struct {
-	message string
-	cause   string
-	series  []mimirpb.LabelAdapter
+func genericValidationError(format string, cause string, series []mimirpb.LabelAdapter) error {
+	return fmt.Errorf(format, cause, formatLabelSet(series))
 }
 
-func (e genericValidationError) Error() string {
-	return fmt.Sprintf(e.message, e.cause, formatLabelSet(e.series))
+func newLabelNameTooLongError(series []mimirpb.LabelAdapter, labelName string) distributorerror.Validation {
+	labelNameTooLongMsgFormat := globalerror.SeriesLabelNameTooLong.MessageWithPerTenantLimitConfig(
+		"received a series whose label name length exceeds the limit, label: '%.200s' series: '%.200s'",
+		maxLabelNameLengthFlag,
+	)
+	return distributorerror.NewValidationError(genericValidationError(labelNameTooLongMsgFormat, labelName, series))
 }
 
-var labelNameTooLongMsgFormat = globalerror.SeriesLabelNameTooLong.MessageWithPerTenantLimitConfig(
-	"received a series whose label name length exceeds the limit, label: '%.200s' series: '%.200s'",
-	maxLabelNameLengthFlag)
-
-func newLabelNameTooLongError(series []mimirpb.LabelAdapter, labelName string) ValidationError {
-	return genericValidationError{
-		message: labelNameTooLongMsgFormat,
-		cause:   labelName,
-		series:  series,
-	}
-}
-
-// labelValueTooLongError is a customized ValidationError, in that the cause and the series are
-// are formatted in different order in Error.
-type labelValueTooLongError struct {
-	labelValue string
-	series     []mimirpb.LabelAdapter
-}
-
-func (e labelValueTooLongError) Error() string {
-	return globalerror.SeriesLabelValueTooLong.MessageWithPerTenantLimitConfig(
-		fmt.Sprintf("received a series whose label value length exceeds the limit, value: '%.200s' (truncated) series: '%.200s'", e.labelValue, formatLabelSet(e.series)),
-		maxLabelValueLengthFlag)
-}
-
-func newLabelValueTooLongError(series []mimirpb.LabelAdapter, labelValue string) ValidationError {
-	return labelValueTooLongError{
-		labelValue: labelValue,
-		series:     series,
-	}
+func newLabelValueTooLongError(series []mimirpb.LabelAdapter, labelValue string) distributorerror.Validation {
+	return distributorerror.NewValidationError(
+		errors.New(
+			globalerror.SeriesLabelValueTooLong.MessageWithPerTenantLimitConfig(
+				fmt.Sprintf("received a series whose label value length exceeds the limit, value: '%.200s' (truncated) series: '%.200s'", labelValue, formatLabelSet(series)),
+				maxLabelValueLengthFlag,
+			),
+		),
+	)
 }
```

Comment:
Now it's more obvious than ever than all this code should live in the `distributor` package.

`distributor` calls `util/validation.ValidateLabels` that calls `distributor/distributorerror` :exploding_head: 

I would stop the scope creep here, and move that in a follow up PR.